packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -14,222 +14,153 @@
+    }
 
-  return res;
-};
-
-/**
- * Convert return type of the service from it's typescript representation to representation in aqua
- * @param returnValue - the value as represented in typescript
- * @param arrowType - the arrow type which describes the service
- * @returns - value represented in aqua
- */
-export const returnType2Aqua = (
-  returnValue: any,
-  arrowType: ArrowType<NonArrowType>,
-) => {
-  if (arrowType.codomain.tag === "nil") {
-    return {};
-  }
-
-  if (arrowType.codomain.items.length === 0) {
-    return {};
+    return value.map((y, i) => {
+      return aqua2ts(y, schema.items[i]);
+    });
+  } else if (["labeledProduct", "struct"].includes(schema.tag)) {
+    if (typeof value !== "object" || value == null || Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return Object.entries(schema.fields).reduce((agg, [key, type]) => {
+      const val = aqua2ts(value[key], type);
+      return { ...agg, [key]: val };
+    }, {});
+  } else {
+    throw new Error("Unexpected tag: " + JSON.stringify(schema));
   }
+}
+
+export function ts2aqua(
+  value: JSONValue,
+  schema: NonArrowSimpleType,
+): JSONValue {
+  if (schema.tag === "nil") {
+    return null;
+  } else if (schema.tag === "option") {
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    return value == null ? [] : ([ts2aqua(value, schema.type)] as [JSONValue]);
+  } else if (
+    schema.tag === "scalar" ||
+    schema.tag === "bottomType" ||
+    schema.tag === "topType"
+  ) {
+    return value;
+  } else if (schema.tag === "array") {
+    if (!Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return value.map((y) => {
+      return ts2aqua(y, schema.type);
+    });
+  } else if (schema.tag === "unlabeledProduct") {
+    if (!Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
 
-  if (arrowType.codomain.items.length === 1) {
-    return ts2aqua(returnValue, arrowType.codomain.items[0]);
+    return value.map((y, i) => {
+      return ts2aqua(y, schema.items[i]);
+    });
+  } else if (["labeledProduct", "struct"].includes(schema.tag)) {
+    if (typeof value !== "object" || value == null || Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return Object.entries(schema.fields).reduce((agg, [key, type]) => {
+      const val = ts2aqua(value[key], type);
+      return { ...agg, [key]: val };
+    }, {});
+  } else {
+    throw new Error("Unexpected tag: " + JSON.stringify(schema));
   }
+}
+
+export const wrapFunction = (
+  value: (...args: JSONValue[]) => JSONValue,
+  schema: ArrowWithoutCallbacks,
+): ((...args: JSONValue[]) => JSONValue) => {
+  return (...args) => {
+    const schemaArgs =
+      schema.codomain.tag === "nil" ? [] : schema.codomain.items;
+
+    const tsArgs = args.map((arg, i) => {
+      return aqua2ts(arg, schemaArgs[i]);
```

Comment:
what if `args` contain more values than `schemaArgs`? Wouldn't it trigger out of bounds exception?